src/network/networkd-dhcp4.c

```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not set DHCPv4 routing policy rule: %m");
+                link_enter_failed(link);
+                return 1;
+        }
+
+        if (link->dhcp4_messages == 0 && !link->network->dhcp_send_decline)
+                dhcp4_check_ready(link);
+        return 1;
+}
+
+static int dhcp4_add_source_routing_policy_rule(Link *link) {
+        struct in_addr address;
+        _cleanup_(routing_policy_rule_freep) RoutingPolicyRule *rule = NULL;
+        _cleanup_free_ char *from_addr_str = NULL;
+        int r;
+
+        assert(link);
+
+        if (!link->dhcp_lease || !link->network) /* link went down while we configured the IP addresses? */
+                return 0;
+
+        if (!link->network->dhcp_source_routing_enabled)
+                return 0;
+
+        r = sd_dhcp_lease_get_address(link->dhcp_lease, &address);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no address: %m");
+
+        r = routing_policy_rule_new(&rule);
+        if (r < 0)
+                return log_oom();
+
+        rule->from.in = address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+
+        /* Only actually add the rule if we didn't already add the same rule from a
+        previous lease. Otherwise, routing_policy_rule_configure will error with -EEXIST */
+        if (
+                link->dhcp4_source_routing_policy_rule &&
+                routing_policy_rule_compare_func(link->dhcp4_source_routing_policy_rule, rule)
+        ) {
+                log_link_debug(link, "New and old source routing rules are the same, won't reconfigure");
+                return 0;
+        }
+        /* Now, if we really are going to set up a new routing policy rule, free the old one.
+        The dhcp4_remove_source_routing_policy_rule function will take care of actually calling
+        routing_policy_rule_remove based on the old lease address later on */
+        if (link->dhcp4_source_routing_policy_rule)
+                routing_policy_rule_free(TAKE_PTR(link->dhcp4_source_routing_policy_rule));
+
+
+        log_link_debug(
+                link, "Adding source routing rule %s/%d -> * table %d pri %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
+
+        r = routing_policy_rule_configure(rule, link, dhcp4_routing_policy_rule_handler);
```

Comment:
Having thought about this for a while, I think rather than just blindly calling `routing_policy_rule_configure`, I should call `routing_policy_rule_get` first, and if it exists, call `routing_policy_rule_make_local`, like here: https://github.com/systemd/systemd/blob/5f04f4e470379f461b9f0f3e4f75e55cfc055536/src/network/networkd-link.c#L910

This is because:
    * `routing_policy_rule_configure` first sends the netlink message to configure the rule, _then_ adds it to its internal set of routing policy rules
    * If the rule already exists in the set, then `routing_policy_rule_configure` fails with `EEXIST`
    * But if that happens, `r <= 0` here is true, and we don't increment `link->dhcp4_messages++`, but
    * The netlink message was already sent, so it'll get a reply. This means that `link->dhcp4_messages` is out of sync with the actual number of outstanding netlink messages, and we're going to fail one of the `assert(link->dhcp4_messages > 0);` statements in one of the handlers

I think it's dangerous that `routing_policy_rule_configure` works this way - it would be better to wrap it in a function that provided its own netlink callback, and only invoked the provided callback if the original call returned success. However, that's probably a job for a different patch, so I'll do the `get` & `make_local` thing here for now? Sound good to you @yuwata ?